File: apps/electron/src/ipc/pdf/ocr.ts
import { ipcMain, BrowserWindow } from "electron";
import { Channels } from "../../constants/channels";
import { goBridge } from "../../services/go-bridge";
import * as fs from "fs";
import * as path from "path";
import * as os from "os";
import * as crypto from "crypto";

// ─── Active OCR job tracking ─────────────────────────────────────────────────
let activeJobId: string | null = null;
let cancelRequested = false;

function getTempDir(): string {
  const dir = path.join(os.tmpdir(), "pdflexity-ocr");
  if (!fs.existsSync(dir)) fs.mkdirSync(dir, { recursive: true });
  return dir;
}

function generateJobId(): string {
  return `ocr-${Date.now()}-${crypto.randomBytes(4).toString("hex")}`;
}

function getMainWindow(): BrowserWindow | null {
  const windows = BrowserWindow.getAllWindows();
  return windows.length > 0 ? windows[0] : null;
}

export function registerOcrHandlers(): void {
  // ── Start OCR processing ─────────────────────────────────────────────────
  ipcMain.handle(
    Channels.PDF_OCR_START,
    async (
      _event,
      buffer: Uint8Array,
      fileName: string,
      languages: string[],
      dpi: number
    ) => {
      const jobId = generateJobId();
      activeJobId = jobId;
      cancelRequested = false;

      const tmpDir = getTempDir();
      const inputPath = path.join(tmpDir, `input-${jobId}.pdf`);
      const outputDir = path.join(tmpDir, jobId);

      try {
        // Write PDF buffer to temp file
        fs.writeFileSync(inputPath, Buffer.from(buffer));
        if (!fs.existsSync(outputDir)) fs.mkdirSync(outputDir, { recursive: true });

        const mainWindow = getMainWindow();

        // Send OCR command to Go bridge with streaming support
        const result = await goBridge.sendStreaming(
          {
            op: "ocr-start",
            inputPath,
            outputPath: outputDir,
            languages,
            dpi,
          },
          (event: any) => {
            // Forward streaming events to the renderer process
            if (!mainWindow || mainWindow.isDestroyed()) return;
            if (cancelRequested && event.type !== "complete" && event.type !== "error") return;

            if (event.type === "progress") {
              mainWindow.webContents.send(Channels.PDF_OCR_PROGRESS, {
                ...event,
                jobId,
              });
            } else if (event.type === "page-result" || event.type === "page-image") {
              mainWindow.webContents.send(Channels.PDF_OCR_PAGE_RESULT, {
                ...event,
                jobId,
              });
            }
          }
        );

        activeJobId = null;

        if (result.success) {
          return { success: true, jobId, data: result.data };
        } else {
          return { success: false, error: result.error || "OCR processing failed" };
        }
      } catch (err: any) {
        activeJobId = null;
        return { success: false, error: err.message || "OCR processing failed" };
      }
    }
  );

  // ── Cancel OCR ───────────────────────────────────────────────────────────
  ipcMain.handle(Channels.PDF_OCR_CANCEL, async (_event, jobId: string) => {
    if (activeJobId === jobId) {
      cancelRequested = true;
      // Send cancel command to Go
      try {
        await goBridge.send({ op: "ocr-cancel" });
        return { success: true };
      } catch {
        return { success: false, error: "Failed to cancel OCR job" };
      }
    }
    return { success: false, error: "No matching active job" };
  });

  // ── Render single page ──────────────────────────────────────────────────
  ipcMain.handle(
    Channels.PDF_OCR_RENDER_PAGE,
    async (_event, buffer: Uint8Array, page: number, scale: number) => {
      const tmpDir = getTempDir();
      const inputPath = path.join(tmpDir, `render-${Date.now()}.pdf`);

      try {
        fs.writeFileSync(inputPath, Buffer.from(buffer));
        const result = await goBridge.send({
          op: "ocr-render-page",
          inputPath,
          page,
          scale: scale || 1.5,
        });

        // Clean up temp file
        try { fs.unlinkSync(inputPath); } catch {}

        if (result.success) {
          return { success: true, data: result.data };
        }
        return { success: false, error: result.error || "Failed to render page" };
      } catch (err: any) {
        try { fs.unlinkSync(inputPath); } catch {}
        return { success: false, error: err.message };
      }
    }
  );

  // ── Export OCR result ───────────────────────────────────────────────────
  ipcMain.handle(
    Channels.PDF_OCR_EXPORT,
    async (
      _event,
      buffer: Uint8Array,
      fileName: string,
      format: string,
      ocrData: any,
      edits: any
    ) => {
      const tmpDir = getTempDir();
      const inputPath = path.join(tmpDir, `export-${Date.now()}.pdf`);
      const baseName = fileName.replace(/\.pdf$/i, "");
      const ext = format === "docx" ? ".docx" : format === "json" ? ".json" : ".pdf";
      const outputPath = path.join(tmpDir, `${baseName}-ocr${ext}`);

      try {
        fs.writeFileSync(inputPath, Buffer.from(buffer));
        
        const result = await goBridge.send({
          op: "ocr-export",
          inputPath,
          outputPath,
          exportFormat: format,
          ocrData: JSON.stringify(ocrData),
          edits: JSON.stringify(edits || {}),
        });

        try { fs.unlinkSync(inputPath); } catch {}

        if (result.success && result.outputPath) {
          const exportedBuffer = fs.readFileSync(result.outputPath);
          const base64 = exportedBuffer.toString("base64");
          try { fs.unlinkSync(result.outputPath); } catch {}
          return {
            success: true,
            data: base64,
            fileName: `${baseName}-ocr${ext}`,
          };
        }
        return { success: false, error: result.error || "Export failed" };
      } catch (err: any) {
        try { fs.unlinkSync(inputPath); } catch {}
        return { success: false, error: err.message };
      }
    }
  );
}

File: src-tauri/src/commands/ocr.rs
//! OCR commands — Rust port of `apps/electron/src/ipc/pdf/ocr.ts`.
//!
//! `pdf_ocr_start` is the only streaming op: the bridge emits
//! `pdf:ocr-progress` / `pdf:ocr-page-result` events as pages are processed and
//! resolves with the terminal summary.

use serde_json::Value;
use std::time::{SystemTime, UNIX_EPOCH};
use tauri::{AppHandle, Manager};

use crate::go_bridge::BridgeHolder;
use crate::go_model::Command;
use crate::result::OpResult;
use crate::util::{cleanup, decode_b64, make_temp_dir, read_file_b64, write_file};

async fn bridge(app: &AppHandle) -> Option<std::sync::Arc<crate::go_bridge::GoBridge>> {
    app.state::<BridgeHolder>().get(app.clone()).await.ok()
}

fn generate_job_id() -> String {
    let nanos = SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .map(|d| d.as_nanos())
        .unwrap_or(0);
    format!("ocr-{nanos:x}")
}

#[tauri::command(rename_all = "camelCase")]
pub async fn pdf_ocr_start(
    buffer_b64: String,
    _file_name: String,
    languages: Vec<String>,
    dpi: i64,
    app: AppHandle,
) -> OpResult {
    let Some(bridge) = bridge(&app).await else {
        return OpResult::err("PDF engine is not available");
    };
    let job_id = generate_job_id();
    let dir = match make_temp_dir("ocr").await {
        Ok(d) => d,
        Err(e) => return OpResult::err(format!("{e:#}")),
    };
    let result = (async {
        let bytes = decode_b64(&buffer_b64)?;
        let input = write_file(&dir, "input.pdf", &bytes).await?;
        let output_dir = dir.join("out");
        tokio::fs::create_dir_all(&output_dir).await?;
        let resp = bridge
            .send_streaming(
                Command {
                    languages: Some(languages),
                    dpi: Some(dpi),
                    ..Command::new("ocr-start")
                        .input_path_opt(input.clone())
                        .output_path_opt(output_dir.clone())
                },
                job_id.clone(),
            )
            .await?;
        if !resp.success {
            anyhow::bail!(resp.error.unwrap_or_else(|| "OCR processing failed".into()));
        }
        Ok::<_, anyhow::Error>(resp.data.unwrap_or(Value::Null))
    })
    .await;
    cleanup(&dir).await;
    match result {
        Ok(data) => OpResult::ok(data).with_job_id(job_id),
        Err(e) => OpResult::err(format!("{e:#}")),
    }
}

#[tauri::command(rename_all = "camelCase")]
pub async fn pdf_ocr_cancel(_job_id: String, app: AppHandle) -> OpResult {
    let Some(bridge) = bridge(&app).await else {
        return OpResult::err("PDF engine is not available");
    };
    let result = async {
        let resp = bridge.send(Command::new("ocr-cancel")).await?;
        if !resp.success {
            anyhow::bail!(resp.error.unwrap_or_else(|| "Failed to cancel OCR job".into()));
        }
        Ok::<_, anyhow::Error>(())
    }
    .await;
    match result {
        Ok(()) => OpResult::ok(Value::Null),
        Err(e) => OpResult::err(format!("{e:#}")),
    }
}

#[tauri::command(rename_all = "camelCase")]
pub async fn pdf_ocr_render_page(buffer_b64: String, page: i64, scale: f64, app: AppHandle) -> OpResult {
    let Some(bridge) = bridge(&app).await else {
        return OpResult::err("PDF engine is not available");
    };
    let dir = match make_temp_dir("ocr-render").await {
        Ok(d) => d,
        Err(e) => return OpResult::err(format!("{e:#}")),
    };
    let result = (async {
        let bytes = decode_b64(&buffer_b64)?;
        let input = write_file(&dir, "render.pdf", &bytes).await?;
        let resp = bridge
            .send(Command {
                page: Some(page),
                scale: Some(scale),
                ..Command::new("ocr-render-page").input_path_opt(input.clone())
            })
            .await?;
        if !resp.success {
            anyhow::bail!(resp.error.unwrap_or_else(|| "Failed to render page".into()));
        }
        Ok::<_, anyhow::Error>(resp.data.unwrap_or(Value::Null))
    })
    .await;
    cleanup(&dir).await;
    match result {
        Ok(data) => OpResult::ok(data),
        Err(e) => OpResult::err(format!("{e:#}")),
    }
}

#[tauri::command(rename_all = "camelCase")]
pub async fn pdf_ocr_export(
    buffer_b64: String,
    file_name: String,
    format: String,
    ocr_data: Value,
    edits: Value,
    app: AppHandle,
) -> OpResult {
    let Some(bridge) = bridge(&app).await else {
        return OpResult::err("PDF engine is not available");
    };
    let dir = match make_temp_dir("ocr-export").await {
        Ok(d) => d,
        Err(e) => return OpResult::err(format!("{e:#}")),
    };
    let result = (async {
        let bytes = decode_b64(&buffer_b64)?;
        let input = write_file(&dir, "export.pdf", &bytes).await?;
        let base_name = file_name.trim_end_matches(".pdf");
        let ext = match format.as_str() {
            "docx" => ".docx",
            "json" => ".json",
            _ => ".pdf",
        };
        let output = dir.join(format!("{base_name}-ocr{ext}"));
        let resp = bridge
            .send(Command {
                export_format: Some(format),
                ocr_data: Some(serde_json::to_string(&ocr_data)?),
                edits: Some(serde_json::to_string(&edits)?),
                ..Command::new("ocr-export").input_path_opt(input.clone()).output_path_opt(output.clone())
            })
            .await?;
        if !resp.success {
            anyhow::bail!(resp.error.unwrap_or_else(|| "Export failed".into()));
        }
        let b64 = read_file_b64(&output).await?;
        Ok::<_, anyhow::Error>(b64)
    })
    .await;
    cleanup(&dir).await;
    match result {
        Ok(b64) => OpResult::ok_file(b64, format!("{}-ocr", file_name.trim_end_matches(".pdf"))),
        Err(e) => OpResult::err(format!("{e:#}")),
    }
}
